library/ecdh.c

```diff
@@ -94,61 +101,195 @@ void mbedtls_ecdh_init( mbedtls_ecdh_context *ctx )
     memset( ctx, 0, sizeof( mbedtls_ecdh_context ) );
 }
 
+static int mbedtls_ecdh_setup_internal( mbedtls_ecdh_context *ctx,
+                                        mbedtls_ecp_group_id grp )
+{
+    int ret;
+#if defined(MBEDTLS_ECDH_LEGACY_CONTEXT)
+    mbedtls_ecdh_context *real_ctx = ctx;
+#else
+    mbedtls_ecdh_context_mbed *real_ctx;
+
+    if( ctx->ctx.mbed_ecdh != NULL )
+        return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+
+    ctx->ctx.mbed_ecdh = (mbedtls_ecdh_context_mbed*) mbedtls_calloc ( 1,
+                            sizeof( mbedtls_ecdh_context_mbed ) );
+
+    if( ctx->ctx.mbed_ecdh == NULL )
+        return( MBEDTLS_ERR_ECDH_ALLOC_FAILED );
+
+    ctx->var = MBEDTLS_ECDH_VARIANT_MBED;
+    ctx->grp = grp;
+
+    real_ctx = ctx->ctx.mbed_ecdh;
+#endif
+
+    ret = mbedtls_ecp_group_load( &real_ctx->grp, grp );
+    if( ret != 0 )
+    {
+        mbedtls_ecdh_free( ctx );
+        return( MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE );
+    }
+
+    return( 0 );
+}
+
 /*
- * Free context
+ * Setup context
  */
-void mbedtls_ecdh_free( mbedtls_ecdh_context *ctx )
+int mbedtls_ecdh_setup( mbedtls_ecdh_context *ctx, mbedtls_ecp_group_id grp )
 {
     if( ctx == NULL )
+        return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+
+#if defined(MBEDTLS_ECDH_LEGACY_CONTEXT)
+    return( mbedtls_ecdh_setup_internal( ctx, grp ) );
+#else
+    switch( grp )
+    {
+        default:
+            return( mbedtls_ecdh_setup_internal( ctx, grp ) );
+    }
+#endif
+}
+
+static void mbedtls_ecdh_free_internal( mbedtls_ecdh_context *ctx )
+{
+#if defined(MBEDTLS_ECDH_LEGACY_CONTEXT)
+    mbedtls_ecdh_context *real_ctx = ctx;
+#else
+    mbedtls_ecdh_context_mbed *real_ctx;
+
+    if( ctx->ctx.mbed_ecdh == NULL )
         return;
 
-    mbedtls_ecp_group_free( &ctx->grp );
-    mbedtls_ecp_point_free( &ctx->Q   );
-    mbedtls_ecp_point_free( &ctx->Qp  );
-    mbedtls_ecp_point_free( &ctx->Vi  );
-    mbedtls_ecp_point_free( &ctx->Vf  );
-    mbedtls_mpi_free( &ctx->d  );
-    mbedtls_mpi_free( &ctx->z  );
-    mbedtls_mpi_free( &ctx->_d );
+    real_ctx = ctx->ctx.mbed_ecdh;
+
+    ctx->var = MBEDTLS_ECDH_VARIANT_NONE;
+    ctx->grp = MBEDTLS_ECP_DP_NONE;
+#endif
+
+    mbedtls_ecp_group_free( &real_ctx->grp );
```

Comment:
I think it would be better to put the code that handles the “MBED” variant context in a separate function, to pave the way for a build that doesn't include this context variant. Make `mbedtls_ecdh_free_internal` take a `mbedtls_ecdh_context_mbed*` as argument. This goes for other functions with a similar structure.

Reply:
Separating them would result in duplicating a lot of code and this means extra maintenance and opportunity for errors. I think that removing the legacy version is pretty straightforward and therefore I would prefer to keep them as they are.